### PR TITLE
[Testcase Refactoring]Make test_sharding_plan.py device-generic and add hw_classification label

### PR DESCRIPTION
**Pull request by youlituao**
## Summary

The three tests in `test/distributed/_shard/sharding_plan/test_sharding_plan.py`
verify `ShardingPlan` / `ShardingPlanner` sharding rules and error handling —
generic distributed logic that is not CUDA-specific. However, the test file
hardcodes CUDA device strings, uses `@requires_nccl()` for admission, and
relies on a hardcoded backend list, preventing out-of-tree accelerator
backends (e.g. PrivateUse1 backends such as NPU) from reusing the tests.

This PR replaces all CUDA-specific and hardcoded-backend patterns with
device-generic equivalents based on the capability registration mechanism
(PR #191919), and adds the `hw_classification = HardwareClassification.ACCELERATOR`
label per the distributed test class labeling rules.

## Changes

- Replaced `@requires_nccl()` with
  `@requires_capabilities(Capability.distributed.backend)` on all three
  test methods. This gates on device-declared capability rather than a
  hardcoded backend name list — new backends only need to override
  `_capabilities()` instead of extending the common backend mapping.
- Changed `@with_comms(init_rpc=False)` to use the default `backend=None`,
  which resolves to `self.backend` at runtime via the
  `ShardedTensorTestBase.backend` property (PR #194992). No module-level
  `backend` or `device_type` globals are needed.
- Added `device_type = torch.device(device).type` at the start of each
  test method, deriving the bare device type (e.g. `"npu"`) from the
  `instantiate_device_type_tests`-injected `device` parameter (e.g.
  `"npu:0"`). This avoids the pitfall where
  `torch.device(device, self.rank)` fails due to duplicate index when
  `device` already carries an index string.
- Replaced `.to(torch.device(device, self.rank))` with `.to(device_type)`
  in `test_sharding_plan_errors` and `test_custom_sharding_planner`.
- Replaced hardcoded `"cuda"` in placement strings with
  `f"rank:2/{device_type}:2"` in `test_shard_module_sub_process_group`
  (4 placement strings).
- Replaced `ChunkAllShardingPlanner(device_type=device)` with
  `ChunkAllShardingPlanner(device_type=device_type)` — passes the bare
  device type string instead of the indexed device string.
- Added `device` parameter to all three test method signatures
  (`test_xxx(self)` → `test_xxx(self, device)`, framework injection).
  The `device` parameter triggers `setUpClass` which resolves
  `privateuse1` to the actual backend name; it is also used to derive
  `device_type` via `torch.device(device).type` at the start of each
  method.
- Replaced `instantiate_device_type_tests(TestShardingPlan, globals())`
  with `instantiate_device_type_tests(TestShardingPlan, globals(),
  except_for="cpu", allow_xpu=True)` — excludes CPU variants at
  instantiation layer (these tests require a distributed accelerator
  backend) and enables XPU variants.

  Note: before this PR these distributed tests never actually executed
  on XPU (`with_comms` skipped them on backend mismatch, `"nccl" !=
  "xccl"`), so `allow_xpu=True` is coverage enablement rather than
  conservation — the tests are device-agnostic after the refactor and
  the dynamic backend resolution path is the same one verified on
  PrivateUse1 (HCCL) and CUDA (NCCL). XPU execution itself will be
  confirmed by XPU CI / follow-up testing.
- Added `hw_classification = HardwareClassification.ACCELERATOR` to
  `TestShardingPlan`.

## Not changed

- `TEST_GPU_NUM` naming retained for backward compatibility (the constant
  itself is hardware-agnostic; renaming is a follow-up).
- `ChunkAllShardingPlanner.__init__` signature unchanged (retains
  `device_type` parameter with default `"cuda"`).
- The numerical assertions and test logic are unchanged.
- `test_common.py` not modified — upstream main already uses
  `self.fc1.to(torch.device(rank))` which correctly resolves to the
  current accelerator device.

## Dependencies

This PR depends on the following infrastructure PRs (all submitted, not
yet merged into upstream main):

| Infrastructure | Source PR | Status |
|---|---|---|
| Capability registration mechanism (`Capability` + `requires_capabilities`) | #191919 | Submitted |
| `PrivateUse1TestBase._capabilities()` override | #193080 | Submitted |
| `ShardedTensorTestBase.backend` property + `with_comms(backend=None)` default | #194992 | Submitted |
| `skip_if_lt_x_gpu` hardware-agnostic implementation | #187650 | Submitted |
| `ShardedTensorTestBase.init_pg` dynamic device binding | #190528 | Submitted |
| `HardwareClassification` enum | #190991 | Merged |

Without #193080, `@requires_capabilities` silently skips PrivateUse1
variants (tests pass as skipped, not failed). This PR can merge before
#193080.

## Test plan

- `python test/distributed/_shard/sharding_plan/test_sharding_plan.py` on CPU
  (no accelerator): no variants generated (`except_for="cpu"`,
  `Ran 0 tests`).
- `python test/distributed/_shard/sharding_plan/test_sharding_plan.py` on an
  out-of-tree accelerator backend (PrivateUse1, torch_npu 2.13.0+git45fbeae
  on PyTorch 2.13.0a0+gitfad7424, Ascend 910B3, 4 NPUs):
  3 NPU variants pass (`ok`), CPU variants not generated (`OK`).
- `python test/distributed/_shard/sharding_plan/test_sharding_plan.py` on CUDA
  (NVIDIA A100-SXM4-80GB x8, PyTorch 2.13.0a0): 3 CUDA variants pass (`ok`),
  CPU variants not generated (`OK`).

### CODE REVIEW

File: test/distributed/_shard/sharding_plan/test_sharding_plan.py [hunk 1]
```diff
@@ -49,10 +60,13 @@ def build_plan(self, module: nn.Module) -> ShardingPlan:
 
 
 class TestShardingPlan(ShardedTensorTestBase):
+    hw_classification = HardwareClassification.ACCELERATOR
+
+    @onlyAccelerator
     @with_comms(init_rpc=False)
```
Comment (JamesD18) [inline test/distributed/_shard/sharding_plan/test_sharding_plan.py:63]: @with_comms(init_rpc=False) needs to be judged via the capability‑registration mechanism.

File: test/distributed/_shard/sharding_plan/test_sharding_plan.py [hunk 2]
```diff
@@ -49,10 +60,13 @@ def build_plan(self, module: nn.Module) -> ShardingPlan:
...
     @skip_if_lt_x_gpu(TEST_GPU_NUM)
-    @requires_nccl()
-    def test_sharding_plan_errors(self):
+    @requires_accelerator_dist_backend(["nccl", "xccl", "privateuse1"])
```
Comment (JamesD18) [inline test/distributed/_shard/sharding_plan/test_sharding_plan.py]: @requires_accelerator_dist_backend(["nccl", "xccl", "privateuse1"]) needs to be judged via the capability‑registration mechanism.

Comment (youlituao) [reply test/distributed/_shard/sharding_plan/test_sharding_plan.py]: Done. Replaced @requires_accelerator_dist_backend(["nccl", "xccl", "privateuse1"]) with @requires_capabilities(Capability.distributed.backend) (PR #191919). This gates on device-declared capability rather than a hardcoded backend name list — new backends only need to override _capabilities() instead of extending the common backend mapping. For PrivateUse1 backends, the capability is registered via PrivateUse1TestBase._capabilities() override (PR #193080). When the capability is not declared, tests are skipped (not failed), so this PR can merge before #193080. Verified on NPU (3 variants pass) and CUDA (3 variants pass, no regression).

File: test/distributed/_shard/sharding_plan/test_sharding_plan.py (hunk 1 above)
Comment (youlituao) [reply test/distributed/_shard/sharding_plan/test_sharding_plan.py:63]: Updated. @with_comms(init_rpc=False) uses the default backend=None, which resolves to self.backend at runtime via the ShardedTensorTestBase.backend property (PR #194992). The backend property resolves the correct distributed backend from self.device_type (set by instantiate_device_type_tests) via dist.get_default_backend_for_device(device_type), with privateuse1 transparently resolved to the actual backend name (e.g. npu) via torch._C._get_privateuse1_backend_name(). No module-level backend or device_type globals are needed. The admission control is handled by @requires_capabilities(Capability.distributed.backend) on each test method, which gates on device-declared capability.

**Review by JamesD18 — APPROVED** [review]
LGTM.Thanks for addressing the review comments.
